File: frontend/src/scene/cando_cylinders.js
/**
 * scene/cando_cylinders.js — the "CanDo style output" cylinder overlay.
 *
 * Draws the FEM-predicted shape exactly the way CanDo does: one tube per helix, a
 * chain of short cylinders (radius = duplex radius) threaded through the per-bp axis
 * positions, plus thin connector cylinders at the crossovers (the "joints").  PURE
 * CYLINDERS — no sphere fillers — so it keeps CanDo's characteristic "coin-stacked"
 * look (one flat disc per base pair) rather than a smooth tube.
 *
 * Each segment is heat-mapped by the RMSF at its base pair with CanDo's jet colour
 * ramp (bluest = rmsf_min / 0th percentile → reddest = rmsf_p95 / 95th percentile,
 * clamped above), matching CanDo's structure_NMA_RMSF.bild.  Segments with no RMSF
 * (or a job run without NMA) fall back to grey.
 *
 * Standalone representation: the caller hides the native NADOC model (via
 * setDesignVisible) so only these tubes show, like the mrDNA CG-beads mode.
 * Display-only / Physical layer — never touches topology.
 *
 * Usage:
 *   const cyl = initCandoCylinders(scene)
 *   cyl.update(data)   // data from GET /cando/jobs/{id}/cylinders
 *   cyl.clear()
 */

import * as THREE from 'three'
import { colormapRGB } from '../ui/colormaps.js'

const _GREY = new THREE.Color(0x8a8a8a)   // fallback when a segment has no RMSF
// Dim the jet ramp: full-saturation jet under an unlit material reads as glaring neon;
// scaling toward ~⅔ brightness keeps the hues but tones it down to CanDo's look.
// Exported so the colour-map legend renders the SAME dimmed jet as the tubes.
export const JET_BRIGHTNESS = 0.62

// Unit cylinder aligned along +Y (height 1), scaled per-instance.
const _CYL_GEO = new THREE.CylinderGeometry(1, 1, 1, 10, 1)
const _Y = new THREE.Vector3(0, 1, 0)
const _a = new THREE.Vector3()
const _b = new THREE.Vector3()
const _mid = new THREE.Vector3()
const _dir = new THREE.Vector3()
const _quat = new THREE.Quaternion()
const _scale = new THREE.Vector3()
const _m = new THREE.Matrix4()
const _col = new THREE.Color()

// CanDo's RMSF heat map is a jet ramp.  We use the VIVID variant (bright blue → cyan →
// green → yellow → red, no dark blue/red tails) so the low-flexibility core reads as a
// bright blue rather than near-black — matching how CanDo's render actually looks (their
// dark ramp tails are lifted by Chimera's lighting; our unlit material would show them raw).
const _JET = [[0, 0, 1], [0, 1, 1], [0, 1, 0], [1, 1, 0], [1, 0, 0]]

/** CanDo-style jet heat-map colour for t∈[0,1] → [r,g,b] in [0,1].
 *  t=0 → blue, 0.25 → cyan, 0.5 → green, 0.75 → yellow, 1 → red. */
export function jetRGB(t) {
  const x = Math.max(0, Math.min(1, Number.isFinite(t) ? t : 0))
  const seg = x * (_JET.length - 1)
  const i = Math.min(_JET.length - 2, Math.floor(seg))
  const f = seg - i
  const a = _JET[i], b = _JET[i + 1]
  return [a[0] + (b[0] - a[0]) * f, a[1] + (b[1] - a[1]) * f, a[2] + (b[2] - a[2]) * f]
}

/**
 * Flatten cylinder data into coloured segments (pure, unit-tested).  Each tube segment
 * joins consecutive axis nodes and carries the MEAN RMSF of its two endpoints; joints
 * carry their own precomputed RMSF.  Returns { tubes, joints } as
 * [{ a:[x,y,z], b:[x,y,z], rmsf: number|null }].
 */
export function cylinderSegments(data) {
  const tubes = []
  for (const h of data?.helices || []) {
    const pts = h.points || []
    const rm = h.rmsf || []
    for (let i = 1; i < pts.length; i++) {
      tubes.push({ a: pts[i - 1], b: pts[i], rmsf: _mean(rm[i - 1], rm[i]) })
    }
  }
  const joints = []
  const jr = data?.joint_rmsf || []
  ;(data?.joints || []).forEach((j, i) => {
    if (Array.isArray(j) && j.length === 2) {
      joints.push({ a: j[0], b: j[1], rmsf: Number.isFinite(jr[i]) ? jr[i] : null })
    }
  })
  return { tubes, joints }
}

function _mean(x, y) {
  const v = [x, y].filter((n) => Number.isFinite(n))
  return v.length ? v.reduce((s, n) => s + n, 0) / v.length : null
}

export function initCandoCylinders(scene) {
  const _meshes = []

  function clear() {
    for (const m of _meshes) {
      scene.remove(m)
      m.material?.dispose?.()
      m.dispose?.()
    }
    _meshes.length = 0
  }

  /** One InstancedMesh of cylinders for `segs` ({a,b,rmsf}); per-instance colormap
   *  colour over [lo,hi] (grey when a segment has no RMSF / no ramp).  `colormap`
   *  names the shared registry ramp (default jet — CanDo's heat map). */
  function _instanced(segs, radius, lo, hi, hasRmsf, colormap = 'jet') {
    if (!segs.length) return
    // Unlit white material tinted PER-INSTANCE via instanceColor (setColorAt) — the jet
    // ramp then reads VIVID like CanDo's flat colormap render, independent of scene
    // lighting.  NB do NOT set vertexColors:true — that looks for geometry vertex colours
    // (the cylinder has none → renders black); instanceColor is the per-instance channel.
    const mat = new THREE.MeshBasicMaterial({ color: 0xffffff })
    const mesh = new THREE.InstancedMesh(_CYL_GEO, mat, segs.length)
    mesh.frustumCulled = false
    const span = hi - lo
    let n = 0
    for (const s of segs) {
      const p = s.a, q = s.b
      if (!p || !q) continue
      _a.set(p[0], p[1], p[2])
      _b.set(q[0], q[1], q[2])
      _dir.subVectors(_b, _a)
      const len = _dir.length()
      if (len < 1e-6) continue
      _dir.divideScalar(len)
      _mid.addVectors(_a, _b).multiplyScalar(0.5)
      _quat.setFromUnitVectors(_Y, _dir)
      _scale.set(radius, len, radius)
      _m.compose(_mid, _quat, _scale)
      mesh.setMatrixAt(n, _m)
      if (hasRmsf && Number.isFinite(s.rmsf) && span > 1e-9) {
        const [r, g, bl] = colormapRGB(colormap, (s.rmsf - lo) / span)
        // treat ramp values as display (sRGB) colours, dimmed from neon toward CanDo's tone
        _col.setRGB(r * JET_BRIGHTNESS, g * JET_BRIGHTNESS, bl * JET_BRIGHTNESS, THREE.SRGBColorSpace)
      } else {
        _col.copy(_GREY)
      }
      mesh.setColorAt(n, _col)
      n++
    }
    mesh.count = n
    mesh.instanceMatrix.needsUpdate = true
    if (mesh.instanceColor) mesh.instanceColor.needsUpdate = true
    scene.add(mesh)
    _meshes.push(mesh)
  }

  let _lastData = null   // cached response so the colour scale / colormap can recolour live

  /** Rebuild the tube overlay from the cached data at [lo,hi] on `colormap`. */
  function _draw(lo, hi, colormap) {
    clear()
    const data = _lastData
    if (!data || (!data.helices?.length && !data.joints?.length)) return
    const { tubes, joints } = cylinderSegments(data)
    const tubeR = Number.isFinite(data.tube_radius_nm) ? data.tube_radius_nm : 1.125
    const jointR = Number.isFinite(data.joint_radius_nm) ? data.joint_radius_nm : 0.2
    const hasRmsf = !!data.has_rmsf
    _instanced(tubes, tubeR, lo, hi, hasRmsf, colormap)
    _instanced(joints, jointR, lo, hi, hasRmsf, colormap)
  }

  return {
    /** Rebuild the tube overlay from a /cando/jobs/{id}/cylinders response.  Optional
     *  {lo, hi, colormap} override the RMSF window + ramp (default data min→p95, jet). */
    update(data, { lo = null, hi = null, colormap = 'jet' } = {}) {
      _lastData = data
      const dlo = Number.isFinite(lo) ? lo : (Number.isFinite(data?.rmsf_min) ? data.rmsf_min : 0)
      const dhi = Number.isFinite(hi) ? hi : (Number.isFinite(data?.rmsf_p95) ? data.rmsf_p95 : 0)
      _draw(dlo, dhi, colormap)
    },
    /** Recolour the current tubes to a new RMSF window / colormap (no re-fetch). */
    recolor(lo, hi, colormap = 'jet') { _draw(lo, hi, colormap) },
    clear,
    active: () => _meshes.length > 0,
  }
}

File: backend/core/cando_cylinders.py
"""CanDo-style "jointed cylinder" geometry for the FEM-predicted shape.

CanDo renders a solved structure as a set of smooth tubes — one per helix, each a
chain of short cylinders (radius = duplex radius) threaded through the per-bp axis
positions — plus thin connector cylinders at the crossovers (the "joints").  This
module reproduces that representation from a CanDo job's cached display positions
(the ``deformed_positions`` list, already rigid-body aligned to the displayed frame),
so the in-app "CanDo style output" toggle draws exactly the familiar CanDo look.

Purely geometric / Physical-layer — display-only, never mutates topology.

The tubes thread through the FEM AXIS nodes (the true helix centre, one per duplex-core
bp) — NOT the backbone midpoint, which precesses around the axis along the helical groove
and makes the tube wobble.  Because the axis nodes exist only for the meshed duplex core,
ssDNA ends are naturally excluded (no grey fall-back).  The aligned axis nodes are cached
with the job (``display.json`` ``axis``); for older jobs cached without them,
:func:`axis_from_backbones` reconstructs a (wobblier) axis from the backbone midpoints,
restricted to the RMSF/duplex-core bp so ssDNA is still excluded.
"""

from __future__ import annotations

from typing import Dict, List, Optional, Tuple

import numpy as np

from backend.core.models import Design
from backend.physics.fem_solver import HELIX_DIAMETER

# Cylinder radii (nm), matching the CanDo BILD: helix tube = duplex radius (11.25 Å),
# crossover joint connector = thin (2.0 Å).
TUBE_RADIUS_NM = HELIX_DIAMETER / 2.0   # 1.125 nm
JOINT_RADIUS_NM = 0.20                   # nm


def axis_from_backbones(
    display_positions: List[dict],
    rmsf: Optional[List[dict]] = None,
) -> List[dict]:
    """Fallback axis for jobs cached WITHOUT the solver's ``axis`` list: reconstruct each
    duplex bp's centre as the midpoint of its two strand backbones (copy 0).  Restricted
    to bp that carry an RMSF node (the meshed duplex core) so ssDNA ends are excluded —
    at the cost of the backbone-groove wobble the cached axis avoids.  Returns the same
    ``[{helix_id, bp_index, position}]`` shape as the solver's axis."""
    rmsf_bp = {(r["helix_id"], r["bp_index"]) for r in rmsf or []}
    fwd: Dict[Tuple[str, int], List[float]] = {}
    rev: Dict[Tuple[str, int], List[float]] = {}
    for p in display_positions:
        if p.get("copy", 0) != 0:
            continue
        key = (p["helix_id"], p["bp_index"])
        if rmsf_bp and key not in rmsf_bp:      # keep only meshed duplex core (drop ssDNA)
            continue
        pos = p.get("backbone_position")
        if pos is None:
            continue
        if p.get("direction") == "FORWARD":
            fwd[key] = pos
        elif p.get("direction") == "REVERSE":
            rev[key] = pos
    out: List[dict] = []
    for key, a in fwd.items():
        b = rev.get(key)
        if b is None:
            continue
        out.append({"helix_id": key[0], "bp_index": key[1],
                    "position": [(a[0] + b[0]) / 2.0, (a[1] + b[1]) / 2.0, (a[2] + b[2]) / 2.0]})
    return out


def compute_cylinders(
    design: Design,
    axis_nodes: List[dict],
    rmsf: Optional[List[dict]] = None,
) -> dict:
    """Build the CanDo-style cylinder representation of a job's predicted shape.

    ``axis_nodes`` is the solver's per-bp helix-CENTRE list ``[{helix_id, bp_index,
    position}]`` (from ``display.json`` ``axis``, or :func:`axis_from_backbones`).  Only
    duplex-core bp appear → no ssDNA, no grey fall-back.

    Returns::

        {"tube_radius_nm": 1.125, "joint_radius_nm": 0.2,
         "helices": [{"helix_id": str,
                      "points": [[x,y,z], ...],   # per-helix axis tube, bp-ordered
                      "rmsf":   [float|None, ...]}],   # per-node RMSF (nm), parallel to points
         "joints":  [[[x,y,z],[x,y,z]], ...],          # crossover connectors
         "joint_rmsf": [float|None, ...],              # per-joint RMSF (mean of both ends)
         "has_rmsf": bool, "rmsf_min": float, "rmsf_p95": float, "rmsf_max": float,
         "n_helices": int, "n_joints": int}

    Positions are in nm in the displayed (aligned) frame — the same frame the deform
    overlay uses — so the tubes drop straight onto the model without a jump.

    When ``rmsf`` (the job's per-node NMA RMSF, ``[{helix_id, bp_index, rmsf_nm}]``) is
    given, every cylinder segment is tagged with the RMSF at its bp so the overlay can
    colour it with CanDo's jet heat map — bluest at the ``rmsf_min`` (0th percentile),
    reddest at the ``rmsf_p95`` (95th percentile, clamped above), exactly as CanDo's
    ``structure_NMA_RMSF.bild`` does.
    """
    axis: Dict[Tuple[str, int], List[float]] = {
        (n["helix_id"], n["bp_index"]): n["position"] for n in axis_nodes
    }
    rmsf_by_node: Dict[Tuple[str, int], float] = {}
    for r in rmsf or []:
        v = r.get("rmsf_nm")
        if v is not None:
            rmsf_by_node[(r["helix_id"], r["bp_index"])] = float(v)

    helices: List[dict] = []
    for helix in design.helices:
        keyed = sorted(
            (bp, axis[(hid, bp)]) for (hid, bp) in axis if hid == helix.id
        )  # sort by bp index → ordered along the helix
        if len(keyed) >= 2:
            helices.append({
                "helix_id": helix.id,
                "points": [p for _, p in keyed],
                "rmsf": [rmsf_by_node.get((helix.id, bp)) for bp, _ in keyed],
            })

    joints: List[list] = []
    joint_rmsf: List[Optional[float]] = []
    for xo in design.crossovers:
        a = axis.get((xo.half_a.helix_id, xo.half_a.index))
        b = axis.get((xo.half_b.helix_id, xo.half_b.index))
        if a is not None and b is not None:
            joints.append([a, b])
            ra = rmsf_by_node.get((xo.half_a.helix_id, xo.half_a.index))
            rb = rmsf_by_node.get((xo.half_b.helix_id, xo.half_b.index))
            both = [x for x in (ra, rb) if x is not None]
            joint_rmsf.append(sum(both) / len(both) if both else None)

    vals = [v for v in rmsf_by_node.values()]
    has_rmsf = len(vals) > 0
    rmsf_min = float(min(vals)) if has_rmsf else 0.0
    rmsf_max = float(max(vals)) if has_rmsf else 0.0
    # 95th percentile → reddest, matching CanDo's HeatMapRange4RMSF (clamp above p95).
    rmsf_p95 = float(np.percentile(vals, 95)) if has_rmsf else 0.0

    return {
        "tube_radius_nm": TUBE_RADIUS_NM,
        "joint_radius_nm": JOINT_RADIUS_NM,
        "helices": helices,
        "joints": joints,
        "joint_rmsf": joint_rmsf,
        "has_rmsf": has_rmsf,
        "rmsf_min": rmsf_min,
        "rmsf_p95": rmsf_p95,
        "rmsf_max": rmsf_max,
        "n_helices": len(helices),
        "n_joints": len(joints),
    }
